### PR TITLE
Updated MoPub Adapter with UnifiedNativeAdMapper Support.

### PR DESCRIPTION
**Pull request by spilaka**
Updated MoPub Adapter with UnifiedNativeAdMapper Support and implemented mediaContentAspectRatio.

### CODE REVIEW

File: ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubAdapter.java [hunk 1]
```diff
@@ -109,9 +109,13 @@ public void requestNativeAd(final Context context,
             listener.onAdFailedToLoad(MoPubAdapter.this, AdRequest.ERROR_CODE_INVALID_REQUEST);
             return;
         }
-
+        if (!mediationAdRequest.isUnifiedNativeAdRequested() &&
+                !mediationAdRequest.isAppInstallAdRequested()) {
+            Log.e(TAG, "Content Ads are not supported.");
```
Comment (chauduyphanvu) [inline ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubAdapter.java]: In addition to saying what's not supported, I'd also log what's supported and the remediation step to be supported.

File: ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubUnifiedNativeAdMapper.java [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+                    break;
+                default:
+                    params.gravity = Gravity.TOP | Gravity.END;
+            }
+            privacyInformationIconImageView.setLayoutParams(params);
+            adView.requestLayout();
+        } else {
+            Log.d(MoPubAdapter.TAG, "Failed to show AdChoices icon.");
+        }
+
```
Comment (chauduyphanvu) [inline ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubUnifiedNativeAdMapper.java]: Remove extra line.

File: ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubUnifiedNativeAdMapper.java [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+                            .withSupportedUrlActions(
+                                    UrlAction.IGNORE_ABOUT_SCHEME,
+                                    UrlAction.OPEN_NATIVE_BROWSER,
+                                    UrlAction.OPEN_IN_APP_BROWSER,
+                                    UrlAction.HANDLE_SHARE_TWEET,
+                                    UrlAction.FOLLOW_DEEP_LINK_WITH_FALLBACK,
+                                    UrlAction.FOLLOW_DEEP_LINK)
+                            .build().handleUrl(context, privacyInformationClickthroughUrl);
+                }
+            });
```
Comment (chauduyphanvu) [inline ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubUnifiedNativeAdMapper.java]: Add new line below.

File: ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubUnifiedNativeAdMapper.java [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+        MoPubNativeMappedImage mainImage = new MoPubNativeMappedImage(
+                nativeAdMainImage, mMoPubNativeAdData.getMainImageUrl(),
+                MoPubAdapter.DEFAULT_MOPUB_IMAGE_SCALE);
+        List<NativeAd.Image> imagesList = new ArrayList<NativeAd.Image>();
+        imagesList.add(mainImage);
+        setImages(imagesList);
+        int height = mainImage.getHeight();
+        int width = mainImage.getWidth();
+        float aspectRatio = 0.0f;
+        if (height != 0) {
```
Comment (chauduyphanvu) [inline ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubUnifiedNativeAdMapper.java]: Could this be a negative int? We might want to check `if (height > 0)`

File: ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubUnifiedNativeAdMapper.java [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+            aspectRatio = (float) (width / height);
+        }
```
Comment (chauduyphanvu) [inline ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubUnifiedNativeAdMapper.java]: Add new line below.

File: ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubUnifiedNativeAdMapper.java [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+    }
+
+    @Override
+    public void recordImpression() {
+    }
+
+    @Override
+    public void handleClick(View view) {
+    }
```
Comment (chauduyphanvu) [inline ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubUnifiedNativeAdMapper.java]: These are not implemented - does Google get notifications by another way?

File: ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubAdapter.java [hunk 7]
```diff
@@ -137,70 +141,71 @@ public void requestNativeAd(final Context context,
...
-                                } catch (MalformedURLException e) {
-                                    Log.d(TAG, "Invalid ad response received from MoPub. Image URLs"
-                                            + " are invalid");
-                                    listener.onAdFailedToLoad(MoPubAdapter.this,
-                                            AdRequest.ERROR_CODE_INTERNAL_ERROR);
+                        if (!(adData instanceof StaticNativeAd)) {
+                            listener.onAdFailedToLoad(MoPubAdapter.this,
+                                    AdRequest.ERROR_CODE_INTERNAL_ERROR);
+                            return;
+                        }
```
Comment (chauduyphanvu) [inline ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubAdapter.java]: Add new line below.

File: ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubAdapter.java [hunk 8]
```diff
@@ -137,70 +141,71 @@ public void requestNativeAd(final Context context,
...
-                                            new URL(staticNativeAd.getMainImageUrl()));
-                                } catch (MalformedURLException e) {
-                                    Log.d(TAG, "Invalid ad response received from MoPub. Image URLs"
-                                            + " are invalid");
-                                    listener.onAdFailedToLoad(MoPubAdapter.this,
-                                            AdRequest.ERROR_CODE_INTERNAL_ERROR);
+                        if (!(adData instanceof StaticNativeAd)) {
+                            listener.onAdFailedToLoad(MoPubAdapter.this,
+                                    AdRequest.ERROR_CODE_INTERNAL_ERROR);
+                            return;
```
Comment (chauduyphanvu) [inline ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubAdapter.java]: Maybe log something here?

File: ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubAdapter.java [hunk 9]
```diff
@@ -137,70 +141,71 @@ public void requestNativeAd(final Context context,
...
+                        }
+                        final StaticNativeAd staticNativeAd = (StaticNativeAd) adData;
+
+                        HashMap<String, URL> map = new HashMap<>();
```
Comment (chauduyphanvu) [inline ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubAdapter.java]: I'd prefer a more descriptive name for this `HashMap`.

File: ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubAdapter.java [hunk 10]
```diff
@@ -137,70 +141,71 @@ public void requestNativeAd(final Context context,
...
+                        try {
+                            map.put(DownloadDrawablesAsync.KEY_ICON,
+                                    new URL(staticNativeAd.getIconImageUrl()));
+                            map.put(KEY_IMAGE,
+                                    new URL(staticNativeAd.getMainImageUrl()));
+                        } catch (MalformedURLException e) {
+                            Log.d(TAG, "Invalid ad response received from MoPub. Image URLs"
+                                    + " are invalid");
```
Comment (chauduyphanvu) [inline ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubAdapter.java]:
```suggestion
                                    + " are malformed.");
```

File: ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubAdapter.java [hunk 11]
```diff
@@ -137,70 +141,71 @@ public void requestNativeAd(final Context context,
...
-                                }).execute(map);
-                            } catch (Exception e) {
-                                Log.d(TAG, "Exception constructing the native ad");
-                                listener.onAdFailedToLoad(
-                                        MoPubAdapter.this, AdRequest.ERROR_CODE_INTERNAL_ERROR);
+                            @Override
+                            public void onDownloadFailure() {
+                                // Failed to download images, send failure callback.
+                                listener.onAdFailedToLoad(MoPubAdapter.this,
+                                        AdRequest.ERROR_CODE_INTERNAL_ERROR);
```
Comment (chauduyphanvu) [inline ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubAdapter.java]: Let's log the failure.

File: ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubNativeAppInstallAdMapper.java [hunk 12]
```diff
@@ -28,50 +30,55 @@
...
+        mediaView.setImageDrawable(nativeAdMainImage);
         setOverrideClickHandling(true);
         setOverrideImpressionRecording(true);
     }
 
     @Override
     public void untrackView(View view) {
         super.untrackView(view);
-        mMopubNativeAdData.clear(view);
+        mMoPubNativeAdData.clear(view);
```
Comment (chauduyphanvu) [inline ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubNativeAppInstallAdMapper.java]: Do we need a`null` check here if it becomes `null` at any time?

File: ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubNativeAppInstallAdMapper.java [hunk 13]
```diff
@@ -82,7 +89,7 @@ public void untrackView(View view) {
 
     public void trackView(View view) {
 
-        mMopubNativeAdData.prepare(view);
+        mMoPubNativeAdData.prepare(view);
```
Comment (chauduyphanvu) [inline ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubNativeAppInstallAdMapper.java]: Do we need a`null` check here if it becomes `null` at any time?

File: ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubNativeAppInstallAdMapper.java [hunk 14]
```diff
@@ -28,50 +30,55 @@
...
+     * Holds privacy icon placement.
+     */
     private int privacyIconPlacement;
+    /**
+     * Holds MoPub privacy information icon.
+     */
     private ImageView privacyInformationIconImageView;
+    /**
+     * The size of MoPub's privacy icon.
+     */
```
Comment (chauduyphanvu) [inline ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubNativeAppInstallAdMapper.java]: The variable names are already self-explanatory that I don't think the comments are needed.

File: ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubUnifiedNativeAdMapper.java [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+     */
+    private int privacyIconPlacement;
+    /**
+     * Holds MoPub privacy information icon.
+     */
+    private ImageView privacyInformationIconImageView;
+    /**
+     * The size of MoPub's privacy icon.
+     */
+    private int mPrivacyIconSize;
```
Comment (chauduyphanvu) [inline ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubUnifiedNativeAdMapper.java]: The variable names are already self-explanatory that I don't think the comments are needed.

File: ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubUnifiedNativeAdMapper.java [hunk 16]
```diff
@@ -0,0 +1,184 @@
...
+        float aspectRatio = 0.0f;
+        if (height != 0) {
+            aspectRatio = (float) (width / height);
+        }
+        setMediaContentAspectRatio(aspectRatio);
+        ImageView mediaView = new ImageView(context);
+        mediaView.setImageDrawable(nativeAdMainImage);
+        setMediaView(mediaView);
+        setOverrideClickHandling(true);
+        setOverrideImpressionRecording(true);
```
Comment (chauduyphanvu) [inline ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubUnifiedNativeAdMapper.java]: Need some new lines in between so it's easier to read.

File: ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubUnifiedNativeAdMapper.java [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+                return;
+            }
+
+            privacyInformationIconImageView = new ImageView(context);
+            String privacyInformationImageUrl =
+                    mMoPubNativeAdData.getPrivacyInformationIconImageUrl();
+            final String privacyInformationClickthroughUrl = mMoPubNativeAdData
+                    .getPrivacyInformationIconClickThroughUrl();
+
+            if (privacyInformationImageUrl == null) {
```
Comment (chauduyphanvu) [inline ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubUnifiedNativeAdMapper.java]: Or empty.

File: ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubUnifiedNativeAdMapper.java [hunk 18]
```diff
@@ -0,0 +1,184 @@
...
+                    break;
+                case NativeAdOptions.ADCHOICES_BOTTOM_LEFT:
+                    params.gravity = Gravity.BOTTOM | Gravity.START;
+                    break;
+                case NativeAdOptions.ADCHOICES_TOP_RIGHT:
+                    params.gravity = Gravity.TOP | Gravity.END;
+                    break;
+                default:
+                    params.gravity = Gravity.TOP | Gravity.END;
+            }
```
Comment (chauduyphanvu) [inline ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubUnifiedNativeAdMapper.java]: Add new line below.

File: ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubUnifiedNativeAdMapper.java [hunk 19]
```diff
@@ -0,0 +1,184 @@
...
+            privacyInformationIconImageView.setLayoutParams(params);
+            adView.requestLayout();
+        } else {
+            Log.d(MoPubAdapter.TAG, "Failed to show AdChoices icon.");
```
Comment (chauduyphanvu) [inline ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubUnifiedNativeAdMapper.java]: Is this log correct? This is in the `else` for the `if (overlayView instanceof FrameLayout) {` check. If so, the code in there is doing more than just processing + showing the AdChoices icon.

Comment (spilaka) [reply ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubUnifiedNativeAdMapper.java]: Yes, the code here only does checking and processing the AdChoicesIcon.

File: ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubUnifiedNativeAdMapper.java (hunk 15 above)
Comment (spilaka) [reply ThirdPartyAdapters/mopub/mopub/src/main/java/com/mopub/mobileads/dfp/adapters/MoPubUnifiedNativeAdMapper.java]: According to Google coding format, comments should be provided to instance variables.

**Pull request merged**